bundles/org.openhab.binding.phc/src/main/java/org/openhab/binding/phc/internal/handler/PHCBridgeHandler.java

```diff
@@ -104,215 +110,405 @@ public void initialize() {
-     * @param buffer
-     */
-    private void processInputStream(byte[] buffer) {
-        List<Byte> result = new ArrayList<Byte>();
-        List<String> repeat = new ArrayList<String>();
-        int pos = 0;
-        messageFragment = null;
-        Byte moduleAddress = null;
+                // not a known module address
+                if (!modules.contains(module)) {
+                    module = -1;
+                    continue;
+                }
 
-        while (pos < buffer.length) {
-            int messageStart = 0;
+                if (logger.isDebugEnabled()) {
+                    logger.debug("get module: {}", PHCHelper.byteToHexString(module));
+                }
 
-            for (byte b : buffer) {
-                if (modules.contains(b)) {
-                    moduleAddress = b;
-                    messageStart = pos++;
-                    break;
+                byte sizeToggle = buffer.get();
+
+                if (logger.isDebugEnabled()) {
+                    logger.debug("get sizeToggle: {}", PHCHelper.byteToHexString(sizeToggle));
                 }
-                pos++;
-            }
 
-            if (moduleAddress == null) {
-                return;
-            }
+                // read length of command and check if makes sense
+                if ((sizeToggle < 1 || sizeToggle > 9) && ((sizeToggle & 0xFF) < 0x81 || (sizeToggle & 0xFF) > 0x89)) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("sizeToggle incorrect");
+                    }
 
-            if (buffer.length - 1 - pos < 3) {
-                messageFragment = new byte[buffer.length - pos + 1];
-                System.arraycopy(buffer, messageStart, messageFragment, 0, messageFragment.length);
-            }
+                    module = sizeToggle;
+                    continue;
+                }
+
+                // read toggle, size and command
+                int size = (sizeToggle & 0x7F);
+                boolean toggle = (sizeToggle & 0x80) == 0x80;
+
+                byte[] command = new byte[size];
+
+                try {
+                    for (int i = 0; i < size; i++) {
+                        command[i] = buffer.get();
+                    }
+
+                    // log command
+                    if (logger.isDebugEnabled()) {
+                        String bin = "";
+                        for (byte b : command) {
+                            bin += PHCHelper.byteToBinaryString(b);
+                        }
+                        logger.debug("command read: {}", bin);
+                    }
+
+                    // read crc
+                    byte crcByte1 = buffer.get();
+                    byte crcByte2 = buffer.get();
+
+                    short crc = (short) (crcByte1 & 0xFF);
+                    crc |= (crcByte2 << 8);
+
+                    if (logger.isTraceEnabled()) {
+                        logger.trace("get crc: {}", (crcByte1 << 8) | crcByte2);
```

Comment:
This operation is very cheap you could consider leaving out the `is*Enabled` check